File: 2025-07-July-GFG-POTD/July 26 - Majority Element II/Solution.py
class Solution:
    def findMajority(self, arr):
        n = len(arr)
        count1 = 0
        count2 = 0
        candidate1 = 0
        candidate2 = 1  # Different from candidate1

        # Step 1: Find potential candidates
        for num in arr:
            if num == candidate1:
                count1 += 1
            elif num == candidate2:
                count2 += 1
            elif count1 == 0:
                candidate1 = num
                count1 = 1
            elif count2 == 0:
                candidate2 = num
                count2 = 1
            else:
                count1 -= 1
                count2 -= 1

        # Step 2: Verify the candidates
        count1 = count2 = 0
        for num in arr:
            if num == candidate1:
                count1 += 1
            elif num == candidate2:
                count2 += 1

        # Step 3: Collect valid candidates
        result = []
        if count1 > n // 3:
            result.append(candidate1)
        if count2 > n // 3:
            result.append(candidate2)

        return sorted(result)

File: 2025-07-July-GFG-POTD/July 26 - Majority Element II/Solution.java
class Solution {
    public ArrayList<Integer> findMajority(int[] arr) {
        int n = arr.length;

        // Step 1: Find potential candidates
        int count1 = 0, count2 = 0, candidate1 = 0, candidate2 = 1; // candidate2 initialized differently to avoid same as candidate1

        for (int num : arr) {
            if (num == candidate1) {
                count1++;
            } else if (num == candidate2) {
                count2++;
            } else if (count1 == 0) {
                candidate1 = num;
                count1 = 1;
            } else if (count2 == 0) {
                candidate2 = num;
                count2 = 1;
            } else {
                count1--;
                count2--;
            }
        }

        // Step 2: Verify the candidates
        count1 = 0;
        count2 = 0;
        for (int num : arr) {
            if (num == candidate1) count1++;
            else if (num == candidate2) count2++;
        }

        // Step 3: Add valid candidates to result
        ArrayList<Integer> result = new ArrayList<>();
        if (count1 > n / 3) result.add(candidate1);
        if (count2 > n / 3) result.add(candidate2);

        Collections.sort(result); // ensure result is in increasing order
        return result;
    }

}
